File: repeat_string_match.py
# -*- coding: utf-8 -*-

#####################################################################
# File Name:  repeat_string_match.py
# Author: shenming
# Created Time: Sun Jan 26 23:42:19 2020
#####################################################################

import os
import sys

from collections import deque

class Solution:
    def repeatStringMatch(self, A: str, B: str) -> int:
        aLen = len(A)
        qA = deque(A)

        aRots = 0
        while aRots < aLen:
            for c1, c2 in zip(qA, B):
                if c1 != c2:
                    break
            else:
                break
            aRots += 1
            qA.rotate(-1)
        else:
            return -1
        bLen = len(B)
        if aLen >= bLen:
            if aRots <= (aLen - bLen):
                return 1
            else:
                return 2
        else:
            for c in B:
                if c != qA[0]:
                    return -1
                qA.rotate(-1)
            res, r = divmod(bLen - aLen + aRots, aLen)
            if (r > 0):
                res += 1
            return res + 1

if __name__ == "__main__":
    s = Solution()
    print(s.repeatStringMatch("aabaa", "aaab"))

File: repeat_string_match.cpp
/*************************************************************************
> File Name:  repeat_string_match.cpp
> Author: shenming
> Created Time: Sun Jan 26 23:34:16 2020
************************************************************************/

#include "common_lc.h"
class Solution {
public:
    int repeatedStringMatch(string A, string B) {
        for (int i = 0; i < A.size(); i++) {
            int count = 1;
            if (A[i] == B[0]) {
                int ii = i;
                int j = 1;
                for (j = 1; j < B.size(); j++) {
                    ii++;
                    if (ii >= A.size()) {
                        count++;
                        ii = ii - A.size();
                    }
                    if (A[ii] == B[j]) {
                        continue;
                    } else {
                        break;
                    }
                }
                if (j == B.size()) {
                    return count;
                }
            }
        }
        return -1;
    }
};

int main()
{
    Solution s;
    int ret = s.repeatedStringMatch("aabaa", "aaab");
    trace(ret);
    return 0;
}
